### ISSUE TITLE
IL2CPP incompatibility of MediaStreamSample

### ISSUE THREAD

**Issue by akihikodaki** [Unity-Technologies/UnityRenderStreaming#134]
MediaStreamSample passes delegates of instance methods to unmanaged code, which is prohibited on IL2CPP. Maybe you can create `System.GCHandle`s referring to such an delegate and pass them with function pointers of delegates of static methods.

**Comment by karasusan** [comment]
Got it. I'll try it fix this one.

**Comment by sindharta** [comment]
Could you please elaborate which part of the code that you meant ?
We tried to build using il2cpp, and it went successfully.

(we can speak Japanese too, if that's better for you :D )

**Comment by akihikodaki** [comment]
It should cause an error at runtime, not at build. It will throw an exception saying:
```
NotSupportedException: IL2CPP does not support marshaling delegates that point to instance methods to native code.
```

**Comment by sindharta** [comment]
I think I see your point here. The problem lies in these lines.

```
        pc2OnIceConnectionChange = new DelegateOnIceConnectionChange(state => { OnIceConnectionChange(pc2, state); });
        pc1OnIceCandidate = new DelegateOnIceCandidate(candidate => { OnIceCandidate(pc1, candidate); });
        pc2OnIceCandidate = new DelegateOnIceCandidate(candidate => { OnIceCandidate(pc1, candidate); });
        pc2Ontrack = new DelegateOnTrack(e => { OnTrack(pc2, e); });
        pc1OnNegotiationNeeded = new DelegateOnNegotiationNeeded(() => { StartCoroutine(Pc1OnNegotiationNeeded()); });

```
These delegates  that point to instance methods (p1, p2 are also are not static), will then be passed to the native code, and then, the il2cpp exception will occur when we execute the code.

Manually creating static functions for each pc, and declaring the pc variables as static, might be one alternative here.

**Comment by akihikodaki** [comment]
That's right.

> Manually creating static functions for each pc, and declaring the pc variables as static, might be one alternative to fix the issue here.

Yes, but the approach automatically requires a singleton and limits flexibility. I believe only one `webrtc::PeerConnectionObserver` is instantiated in most use case so it may not be a problematic, but it may be in minor use case such as testing. The limitation may also be quite a disturbing if you provide delegates for something else than `webrtc::PeerConnectionObserver`.

**Comment by sindharta** [comment]
I see. Thanks for the hints. We'll see what we can do.

**Comment by Virendrasinh28** [comment]
Is it supported in android(apk)?

**Comment by akihikodaki** [comment]
I do not think so. This library currently lacks:
- Initialization code for Android specifics
- Build infrastructure for Android

They are irrelevant from this issue anyway. You can use this library with Mono once these two are prepared.

**Comment by sindharta** [comment]
@Virendrasinh28 UnityRenderStreaming is using DirectX to stream from PC to browsers, so the answer would be no.

**Comment by karasusan** [comment]
Currently, `IL2CPP` is not supported by our webrtc package.
Of course, we should support it but this change is so huge.

**Comment by karasusan** [comment]
I am fixing this issue on this branch.
I can do it soon.

https://github.com/Unity-Technologies/com.unity.webrtc/tree/fix/il2cpp

**Comment by akihikodaki** [comment]
That is great. It seems fine for me.

I noticed just one thing suspicious:
https://github.com/Unity-Technologies/com.unity.webrtc/commit/75bda4d2202f53b8b3052f223fd527c611a8f388#diff-29e96fc584da4309e648870d72ff1b65R100-R117

A finalizer is added to `Unity.WebRTC.RTCDataChannel`. Finalization is a useful feature but may impact the performance and also be dangerous because it resurrects the object once garbage collected.

You may still implement a finalizer, but Microsoft suggests to implement and use `System.IDisposable` if possible, and suppress the finalizer with `System.GC.SuppressFinalize` in `Dispose` method. It is about .NET Framework, but I'm sure it safely applies to Mono and Unity as it is quite a common problem of garbage collecting.
See https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose?view=netframework-4.8#the-dispose-overload for details.

I did not see anything else regressing. It is a fine work anyway.

**Comment by karasusan** [comment]
Awesome!
I have never known `GC.SuppressFinalize` method to suppress calling finalize method.
I'll fix the dispose logic. Thanks a lot!

**Comment by karasusan** [comment]
Fixed the issue on the `1.1.0` branch.
https://github.com/Unity-Technologies/com.unity.webrtc/blob/release/1.1.0/Runtime/Scripts/MediaStream.cs#L60
